velox/expression/Expr.cpp

```diff
@@ -1128,7 +1128,17 @@ void Expr::addNulls(
   if (result->size() < rows.end()) {
     BaseVector::ensureWritable(
         SelectivityVector::empty(), type, context.pool(), result);
-    result->resize(rows.end());
+
+    // Avoid calling resize on all children by adding top level nulls only.
+    if (result->typeKind() == TypeKind::ROW) {
+      auto oldSize = result->size();
+
+      result->BaseVector::resize(rows.end());
```

Comment:
using BaseVector::resize() here seems risky as the vectors internal implementation details spill over into expr.cpp, is there anyway we can encapsulate this action in a new Vector API?

Additionally, addNulls() has other opportunities for improving perf:
- calling ensureWritable on RowVector can be expensive if the children are shared, but adding nulls does not affect them
- Other vector types can also have some buffers or child vectors that are shared, so we dont necessarily need to call ensureWritable on them too and only make sure the top level nulls buffer is unique

I was wondering if it would make sense to to come up with APIs that allow us to support these use-cases and tackle this perf problem more holistically? Like  ensureNullsWritable() and appendNulls()

Reply:
+1

Would it make sense to add an API to increase RowVector size by adding nulls?

Reply:
make sense i can do that

Reply:
@mbasmanova  I updated the diff by adding appendNulls API to row vector. 
@bikramSingh91  I agree that there several other optimizations to add (avoid calling ensnrewriable for all complex type and wrap in dictionary instead for example when the input vector is not unique)

but this optimization are orthogonal and can be addressed in a different diff.
also there are many similarities between this and nullOutErrors